src/index.js

```diff
@@ -29,17 +29,28 @@ class MediaQuery extends React.Component {
     values: {}
   }
 
-  state = { matches: false }
+  constructor(props) {
+    super(props)
+    this.initMql(props)
+    this.state = {
+      matches: this._mql.matches
+    }
+  }
 
-  componentWillMount() {
+
+  componentDidUpdate(_, prevState) {
+    if(this.state.matches !== prevState.matches) {
+      this.props.onBeforeChange && this.props.onBeforeChange(this.state.matches)
+      this.props.onChange && this.props.onChange(this.state.matches)
+    }
     this.updateQuery(this.props)
```

Comment:
The query should only be updated if the props change, this will update it on every transition between match states.